### PR TITLE
1759 restrict future date selection

### LINKED ISSUE #1759: Restrict Future Date Selections on Date Picker
### Overview
We need an error message or UI feedback to indicate that a user cannot select a date in the future so that we can prevent future date selections on the date picker.

### Background
Currently, when selecting an end date in the date picker calendar, the data load as though data are available for the whole timespan. Additionally, when selecting start and end dates in the future, the loading modal briefly appears. 

### Action Items
- [x] Determine how to limit date picker
- [x] Possibly create error message with either of the following text: "You cannot select a future date. Please choose a date on or before today's date," OR "Error: Date cannot be in the future."

### Resources/Instructions
<details><summary>Loading modal with future end date selected</summary>
<p>
Screen Shot 2024-06-10 at 1 30 57 PM

### PR DESCRIPTION
**Pull request by bphan002**
Fixes #1759

### Issue
Users can select past the current date and still have the page render data. 

### Changes

Future months/days are not possible from the current date since they are not selectable.

  - [x] Up to date with `main` branch
  - [x] Branch name follows [guidelines](https://github.com/hackforla/311-data/blob/master/GETTING_STARTED.md#feature-branching)
  - [x] All PR Status checks are successful
  - [x] Peer reviewed and approved

### CODE REVIEW

**Review by aqandrew — CHANGES_REQUESTED** [review]
1. The error message popup causes a layout shift, which makes for a disruptive user experience:

   https://github.com/hackforla/311-data/assets/9038965/1ef9b91f-19f4-434a-9cef-a4a2dadacdc6

2. The above screen recording shows that if a future date is selected while a date range is selected, the end date gets deselected.

### Possible solution

I'm not sure if there is a suitable place in our UI for such an error message. But we can avoid these 2 problems altogether if we take a different approach to restricting future date selections. Imo it'd be a good idea to follow Google Flights' example, which simply makes past dates non-interactable:

https://github.com/hackforla/311-data/assets/9038965/fc33da6d-c261-45a0-9b59-74d509151666

In other words, no state changes occur if an invalid date is clicked. If we really want to be more explicit, we could also set `cursor: not-allowed` as a hover style for future dates.

What do you think @bberhane? / Was there any input from Design on this?

**Comment by Skydodle** [comment]
Hi @bphan002, sorry I was gonna comment on this before you started working but totally forgot. I actually worked on React DayPicker previously on an issue very similar to this one, and there's a 4 lines solution for this issue that will do everything we wanted.

### Disable the future month navigation
React DayPicker actually has a built-in props for preventing future months navigation. Similar to PR #1686 where we restricted month navigation for unavailable previous years using the `fromMonth` prop, for restricting future months navigation we can use the `toMonth` prop, and pass in the current year and current month:
````javascript
  const currentYear = new Date(today.getFullYear());
  const currentMonth = new Date(today.getMonth());
  
  return (
    <>
      {/* <Styles range={range} /> */}
      <DayPicker
        // other existing props...
        fromMonth={new Date(2019, 12)} // this was from PR#1686 that restricts previous months where our data is not avail
        toMonth={new Date(currentYear, currentMonth)}
      />
    </>
  );
````
This will disable the forward triangle on the top right of the calendar when the calendar is on the current month.
<details><summary>screenshot no forward triangle</summary>
<p>

Screenshot 2024-06-19 at 11 15 02 PM


</p>
</details> 

---

### Disable the future day click events
There's also a built-in prop that disable days, which we've already use it to disable days after today:
````javascript
      <DayPicker
        disabledDays={{ before: lastThreeMonths, after: today }} // existing code
````
But for some reason React DayPicker created this prop to disable days only to be styled with CSS, the colors of disabled days just darkens, it does not disable click event 🤔. It allows you to define what days are disabled, but you need to DIY on what you want to do to these disabled days.

Luckily there is a simple one line solution for this. We already have a CSS modifier class that targets the disabled state of the `DayPicker-Day` element. Similar to what @aqandrew suggested, we can use `pointer-events: none` to achieve the disable mouse event behavior:

````javascipt
    /* Disabled cell */

    '& .DayPicker-Day--disabled': {        // existing styling line 50-55
      color: colors.textSecondaryDark,     // existing styling
      pointerEvents: 'none !important',   // <------ add pointer-events: none
    },
````
This will disable the future days remaining in the current month from being clickable.

Reference: [React DayPicker Docs](https://react-day-picker-v7.netlify.app/api/DayPicker), [MDN pointer events](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events)

**Comment by bphan002** [comment]
@aqandrew Thanks for pointing out that this shifts the layout and is disruptive. @Skydodle thanks for clarifying why the disabledDays prop needed css logic in order to work.  I was trying to figure out why the prop wasn't enough when I was working on it.

I think this is sufficient unless the design team wants to implement an error message for this issue. 

Thank you!

File: components/common/ReactDayPicker/ReactDayPicker.jsx [hunk 1]
```diff
@@ -51,6 +51,7 @@ const useStyles = makeStyles(theme => ({
 
     '& .DayPicker-Day--disabled': {
       color: colors.textSecondaryDark,
+      pointerEvents: 'none !important',
```
Comment (aqandrew) [inline components/common/ReactDayPicker/ReactDayPicker.jsx:54]:
```suggestion
      pointerEvents: 'none',
```

Pretty sweet that resolving this issue boils down to just a single CSS rule! `!important` isn't necessary here though; nothing else sets `pointer-events` on this element

**Review by aqandrew — CHANGES_REQUESTED** [review]
Looking good. I left a comment with just one small change requested, because it's good practice to avoid `!important` where possible

**Review by aqandrew — APPROVED** [review]

**Review by Skydodle — CHANGES_REQUESTED** [review]
@bphan002 I would also disable forward month navigation beyond the current month. This would eliminate any confusion for the users in addition to disabling click event for days. There is no reason to provide users the ability to navigate to future months when they can't click on the days.

**Comment by bphan002** [comment]
Good catch,  I went ahead and did that as well.

**Review by Skydodle — APPROVED** [review]
Tested with no issues. No month navigation beyond current month, and no click event available for future days. Thanks @bphan002 !
